### DISCUSSION TITLE
Unable to Call /Account/Login in Serene ASP.NET Core

### DISCUSSION THREAD

**Discussion by glenngsanchez** [serenity-is/Serenity#5343] [Help] [answered]
Hi All!

I started using Serene ASP.NET Core and trying to call Login API from external application. Login is successful when calling Serene ASP.NET MVC, But this doesn't work when calling Serene ASP.NET Core /Account/Login. I always get 404 Bad Request.

POST http://localhost:59953/Account/Login http/1.1
content-type: application/json

{
"Username": "admin",
"Password": "serene"
}

Please help/point me to the right solution.

Thanks!

**Comment by deveracs** [comment, marked as answer]
Hi glennsanchez, first get **CSRF-TOKEN=xxxxxxx; path=/** in a previous call to http://localhost:59953 and send it in the login /Account/Login header as X-CSRF-TOKEN

I hope it helps you.

**Reply by glenngsanchez** [reply]
hi @deveracs , thanks for reply.  I have successfully make it work. Thanks for the help!

`var getresponse = client.GetAsync(baseurl).Result;
                string csrf = getresponse.Headers.GetValues("Set-Cookie").ToList()[1];
                string token = csrf.Split(";")[0].Split("CSRF-TOKEN=")[1];
                
                var request = new HttpRequestMessage
                {
                    RequestUri = new Uri(baseurl),
                    Method = HttpMethod.Post,
                    Headers = {
                        { "X-CSRF-TOKEN", token }
                    },
                    Content = new StringContent( // Just example of request sending multipart request
                            JsonConvert.SerializeObject(credential),
                            Encoding.UTF8,
                            "application/json")
                };`

**Reply by JohnRanger** [reply]
Hi @deveracs , hi @glenngsanchez 

I am having similar situation like @glenngsanchez (I want to auto-login into a serene .NET5 project)  but I don't understand how to get the CSRF token on client-side as the code above from glennsanchez for getting the token looks like C# code and not javascript/typescript code.

I have to admit that I don't understand the concept of this CSRF token - especially not where it gets created (client-side, server-side) and when. 

My goal is to do an auto-login at the login page (possibly hidding the username + password field) with a given default user / default password as I want to create an electron app with a serene project and do not need any user-specific authentication. 

May I kindly ask to quickly explain me the CSRF thing used within Serene - or pointing me to a documentation where I can better understand this. A quick google didn't reveal anything useful related to CSRF and Serenity framework.

Thanks for any response and

with kind regards,

John

**Reply by deveracs** [reply]
hi @JohnRanger, CSRF token is generated in server-side at first request. In the previous post (@glenngsanchez), the client is another server-side application (C# code). In your case the client is a client-side application (js) and it is the same way. First call ajax request to main page by get, get CSRF token in response, save it in your application and use it in next request (login or another endpoint). 

**Postman** is a great tool to test your request, buid body and headers and inspect the response.

[here a detailed explanation](https://dev.to/_smellycode/csrf-in-action-21n3) 

I hope it helps you.

**Reply by JohnRanger** [reply]
Hi @deveracs , 
many thanks for your feedback. I see, I have some homeworks to do :-)

Have a nice day :-)

With kind regards,

John

**Comment by juliermefelix** [comment]
Dear,
I'm wanting to log in to appgyver using the serenity credentials, but I'm not able to, could you help me by telling me how you managed to use asp.net MCV

![image](https://user-images.githubusercontent.com/7516443/175610917-3339f91d-a4d8-4668-a3da-c8a3d647c657.png)
